### PR TITLE
Add permission guard to midddleware

### PR DESCRIPTION
**Pull request by mr-bjerre**
Add new feature requested in #103

### CODE REVIEW

File: src/nextjs/middleware/withMiddlewareAuth.ts [hunk 1]
```diff
@@ -119,12 +131,20 @@ export const withMiddlewareAuth: withMiddlewareAuth =
...
+        throw new NoPermissionError('User is not permitted, redirecting');
       }
 
       // Authentication successful, forward request to protected route
       return res;
     } catch (err: unknown) {
-      const { redirectTo = '/' } = options;
+      let { redirectTo = '/' } = options;
+      if (err instanceof NoPermissionError && !!options.permissionGuard) {
+        redirectTo = options.permissionGuard.redirectTo;
```
Comment (ferpintado) [inline src/nextjs/middleware/withMiddlewareAuth.ts:146]: I'd default to `redirectTo` from options if this is not defined.

Comment (mr-bjerre) [reply src/nextjs/middleware/withMiddlewareAuth.ts:146]: Yeah good idea. I've added that now

**Review by ferpintado — APPROVED** [review]

**Review by ferpintado — APPROVED** [review]

**Comment by silentworks** [comment]
This PR looks specific to your use case and probably won't work for general use. I can see you are referencing an `admin` table which would be user created.

**Comment by mr-bjerre** [comment]
> This PR looks specific to your use case and probably won't work for general use. I can see you are referencing an `admin` table which would be user created. 

No this is not specific to my use case. I just added the admin table as an example. It could just as well be `user => user.email.endswith("mycompany.com")`.

I'd argue that this solves a _very_ common use case. Ie. making it possible to restrict access to certain roles, users etc. Instead of a binary response - logged in or not. With this you have role based access via middleware.

**Comment by ferpintado** [comment]
I do find it useful as well. In my case, I need to check `user.user_metadata.role` to know whether the user has permissions. 

I do agree, having the example in `examples/nextjs/pages/middleware-protected/_middleware.ts` is very specific. Perhaps just needs to be added in the README.

**Comment by mr-bjerre** [comment]
I'll update example and docs later tonight

**Comment by silentworks** [comment]
@mr-bjerre I know this is a big ask, but can you make this change to the new `main` branch, please?

**Comment by mr-bjerre** [comment]
> @mr-bjerre I know this is a big ask, but can you make this change to the new `main` branch, please?

Opened new PR #126 in favor of this

**Pull request closed without merging**